pkg/api/defaults.go

```diff
@@ -476,23 +488,39 @@ func (p *Properties) setAgentProfileDefaults(isUpgrade, isScale bool) {
 			if profile.Distro == "" {
 				if p.OrchestratorProfile.IsKubernetes() {
 					if profile.OSDiskSizeGB != 0 && profile.OSDiskSizeGB < VHDDiskSizeAKS {
-						profile.Distro = Ubuntu
+						if p.OrchestratorProfile.IsAzureCNI() {
+							// Workaround for https://github.com/Azure/aks-engine/issues/761.
+							profile.Distro = Ubuntu
+						} else {
+							profile.Distro = Ubuntu1804
+						}
 					} else {
-						profile.Distro = AKS
+						profile.Distro = AKS1804
 					}
 				} else {
 					profile.Distro = Ubuntu
 				}
-				// Ensure distro is set properly for N Series SKUs, because
+				// Ensure deprecated distros are overridden
 				// Previous versions of aks-engine required the docker-engine distro for N series vms,
-				// so we need to hard override it in order to produce a working cluster in upgrade/scale contexts
+				// so we need to hard override it in order to produce a working cluster in upgrade/scale contexts.
 			} else if p.OrchestratorProfile.IsKubernetes() && (isUpgrade || isScale) {
-				if profile.Distro == AKSDockerEngine {
-					profile.Distro = AKS
+				if profile.Distro == AKSDockerEngine || profile.Distro == AKS {
+					profile.Distro = AKS1604
 				}
 			}
 		}
 
+		// Override the --resolv-conf kubelet config value for Ubuntu 18.04 after the distro value is set.
+		if profile.IsUbuntu1804() {
+			if profile.KubernetesConfig == nil {
+				profile.KubernetesConfig = &KubernetesConfig{}
+			}
+			if profile.KubernetesConfig.KubeletConfig == nil {
+				profile.KubernetesConfig.KubeletConfig = map[string]string{}
+			}
+			profile.KubernetesConfig.KubeletConfig["--resolv-conf"] = "/run/systemd/resolve/resolv.conf"
```

Comment:
Is there no way to ensure that we set the distro before we set the kubelet defaults? That way we can move this default enforcement into the "set default kubelets configuration" flow.

Reply:
Not really. The kubelet defaults are part of `orchestratorProfile` defaults, which is before `masterProfile` / `agentPoolProfile` defaults. Changing that order seems like a riskier and more complex change to me than moving the kubelet distro-specific override to the profile defaults.

Reply:
If this is really a 16.04- and 18.04-specific setting, then we should also set the value for 16.04. We shouldn't assume some prior default enforcement has already been set with 16.04 defaults in mind.

Reply:
Do you mean explicitly set it to the default (`"/etc/resolv.conf"`) for Ubuntu 16.04? This isn't really a 16.04-/18.04 setting, it's more of a workaround to fix container DNS that is broken with 18.04 due to a breaking change: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1624320/comments/8

Reply:
Btw this isn't a new change, I just moved it after finding out that kubelet defaults were set before the default distro was set. The original change was https://github.com/Azure/aks-engine/pull/223/files#r265284630

Reply:
Got it. So basically it's 18.04-only (as of now) where we have to actually make this change.

Reply:
For consistency, can we also move the equivalent master profile distro-dependent config into the master profile defaults flow and out of the kubelet flow? At least then we have a consistent enforcement pattern when we reconcile kubeletConfig + nodepool.

Reply:
that's already done, unless I'm misunderstanding

Reply:
https://github.com/Azure/aks-engine/pull/1223/files#diff-fd43f080184ed15731e5d7e0d917d6e1R325